app/src/main/java/net/programmierecke/radiodroid2/alarm/AlarmReceiver.java

```diff
@@ -215,6 +217,83 @@ protected void onPostExecute(String result) {
         }.execute();
     }
 
+    private void setAlarmVolume(float volume) {
+        try {
+            if (BuildConfig.DEBUG) { Log.d(TAG, "Setting volume to "+volume); }
+            itsPlayerService.SetVolume(volume);
+        } catch (Exception e) {
+            Log.e(TAG, "Couldn't set volume "+e);
+        }
+    }
+
+    private void graduallyIncreaseAlarmVolume(Context context) {
+        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
+        int slowWakeMillis = sharedPref.getInt("gradually_increase_volume", 0) * 10000;
+
+        if (slowWakeMillis == 0) {
+            if (BuildConfig.DEBUG) { Log.d(TAG, "Gradual alarm volume disabled"); }
+            return;
+        }
+
+        float maxVolume = 100f;
+        float minVolume = 0f;
+        float volumeRange = maxVolume - minVolume;
+        long triggerMillis = System.currentTimeMillis();
+        int streamReadTimeout = sharedPref.getInt("stream_read_timeout", 10) * 10000;
+
+        setAlarmVolume(minVolume);
+
+        Handler handler = new Handler();
+        Runnable runnable = new Runnable() {
+            boolean isPlaying = false;
+            long playStatedMillis = 0;
+            float currentVolume = minVolume;
+            @Override
+            public void run() {
+                if(BuildConfig.DEBUG) { Log.d(TAG, "Increase volume loop"); }
+
+                long totalElapsedMillis = System.currentTimeMillis() - triggerMillis;
+
+                try {
+                    isPlaying = itsPlayerService.isPlaying();
```

Comment:
`isPlaying` is slightly deceiving, it includes pre-play state which includes connecting to server. We should count only the time something is really playing. You can add method to get actual state of the playback from the service.